File: VIKINGYFY-packages/luci-app-wolplus/luasrc/model/cbi/wolplus.lua
local LUCI_SYS = require("luci.sys")

local t = Map("wolplus", translate("Wake on LAN +"), translate("Wake on LAN + is a mechanism to remotely boot computers in the local network.") .. [[<br/><br/><a href="https://github.com/sundaqiang/openwrt-packages" target="_blank">Powered by sundaqiang</a>]])
t.template = "wolplus/index"

local e = t:section(TypedSection, "macclient", translate("Host Clients"))
e.template = "cbi/tblsection"
e.anonymous = true
e.addremove = true

-- Add device section
local a = e:option(Value, "name", translate("Name"))
a.optional = false

-- MAC address
local nolimit_mac = e:option(Value, "macaddr", translate("MAC Address"))
nolimit_mac.rmempty = false
LUCI_SYS.net.mac_hints(function(mac, name)
	nolimit_mac:value(mac, "%s (%s)" % {mac, name})
end)

-- Network interface
local nolimit_eth = e:option(Value, "maceth", translate("Network Interface"))
nolimit_eth.rmempty = false
nolimit_eth.default = "br-lan"
for _, device in ipairs(LUCI_SYS.net.devices()) do
	if device ~= "lo" then
		nolimit_eth:value(device)
	end
end

-- Wake device
local btn = e:option(Button, "_awake", translate("Wake Up Host"))
btn.inputtitle = translate("Awake")
btn.inputstyle = "apply"
btn.disabled = false
btn.template = "wolplus/awake"

-- Generate UUID
local function gen_uuid()
	local uuid = LUCI_SYS.exec("echo -n $(cat /proc/sys/kernel/random/uuid)")
	return uuid:gsub("-", "")
end

-- Create function
function e.create(_, uuid)
	uuid = gen_uuid()
	TypedSection.create(e, uuid)
end

return t

File: VIKINGYFY-packages/luci-app-wolplus/htdocs/luci-static/resources/view/wolplus.js
'use strict';
'require form';
'require rpc';
'require ui';
'require uci';
'require view';
'require tools.widgets as widgets';

var callWake = rpc.declare({
	object: 'luci.wolplus',
	method: 'wake',
	params: [ 'iface', 'mac' ],
	expect: { '': {} }
});

var callHostHints = rpc.declare({
	object: 'luci-rpc',
	method: 'getHostHints',
	expect: { '': {} }
});

function addHostHints(option, hosts) {
	L.sortedKeys(hosts).forEach(function(mac) {
		var hint = hosts[mac].name ||
			L.toArray(hosts[mac].ipaddrs || hosts[mac].ipv4)[0] ||
			L.toArray(hosts[mac].ip6addrs || hosts[mac].ipv6)[0];

		option.value(mac, hint ? '%s (%s)'.format(mac, hint) : mac);
	});

	return option;
}

return view.extend({
	load: function() {
		return Promise.all([
			L.resolveDefault(callHostHints(), {}),
			uci.load('wolplus')
		]);
	},

	render: function(data) {
		var hosts = data[0] || {};
		var m, s, o;
		var view = this;

		m = new form.Map('wolplus', _('Wake on LAN +'),
			_('Wake on LAN + is a mechanism to remotely boot computers in the local network.'));

		s = m.section(form.GridSection, 'macclient', _('Host Clients'));
		s.anonymous = true;
		s.addremove = true;
		s.sortable = true;
		s.nodescriptions = true;

		o = s.option(form.Value, 'name', _('Name'));
		o.rmempty = false;

		o = s.option(form.Value, 'macaddr', _('MAC Address'));
		o.rmempty = false;
		o.datatype = 'macaddr';
		addHostHints(o, hosts);

		o = s.option(widgets.DeviceSelect, 'maceth', _('Network Interface'));
		o.rmempty = false;
		o.default = 'br-lan';
		o.noaliases = true;
		o.noinactive = true;

		s.renderRowActions = function(section_id) {
			var defaultButtons = form.GridSection.prototype.renderRowActions.call(this, section_id);
			var buttonContainer = defaultButtons.querySelector('div');
			var wakeButton = E('button', {
				'type': 'button',
				'class': 'btn cbi-button cbi-button-action',
				'click': ui.createHandlerFn(this, function() {
					return view.handleWakeup(section_id);
				})
			}, _('Awake'));

			if (buttonContainer) {
				var editButton = buttonContainer.querySelector('.cbi-button-edit');
				buttonContainer.insertBefore(wakeButton, editButton || buttonContainer.firstChild);
			}

			return defaultButtons;
		};

		return m.render();
	},

	handleWakeup: function(section_id) {
		var name = uci.get('wolplus', section_id, 'name') || section_id;
		var mac = uci.get('wolplus', section_id, 'macaddr');
		var iface = uci.get('wolplus', section_id, 'maceth') || 'br-lan';

		if (!mac) {
			ui.addNotification(null, E('p', _('Please save the client before waking it up.')), 'error');
			return Promise.resolve();
		}

		return callWake(iface, mac).then(function(res) {
			res = res || {};

			var output = (res.stdout || res.stderr || '').trim();
			var code = res.code;
			var message = output || _('Wake command completed with code %d.').format(code || 0);
			var level = (code == null || code === 0) ? 'info' : 'error';

			ui.addNotification(null, E('p', [
				_('Wake Up Host'), ': ', name, ' (', mac, ')',
				E('br'), message
			]), level);
		}).catch(function(err) {
			ui.addNotification(null, E('p', _('Wake command failed: %s').format(err.message || err)), 'error');
		});
	}
});
